### ISSUE TITLE
Hacking a serial module

### ISSUE THREAD

**Issue by jayepolo** [roy-bentley/enviracom#1]
Honeywell discontinued the serial interface module to connect RS232 to the EnviraCOM bus.  The EnviraCOM specs for the 3 wires are: 2 wires for 24VAC and 1 wire for data.  Can you share what voltage the data line uses in order to hack an alternative hardware interface?

**Comment by roy-bentley** [comment]
Hi,
Everything I know about the module is in [this](https://github.com/roy-bentley/enviracom/blob/28846110eb027c3927404be8edef35fdc40c4ce3/docs/reference/W8735A1005_SPECS.pdf) document. I’ve unfortunately not had time to attempt to build one of these interfaces yet.

**Comment by jgrisham** [comment]
While I haven’t tried this, it may just be easier to connect something like an esp8266 or RPi Zero W _directly_ to debug or IC pins inside a working EnviroCOM thermostat, equipment module, sensor, or zone module… that way, the existing circuitry will handle the _(AC voltage? … edit: nope, it’s DC; see below)_ physical layer _(possibly including bus arbitration and collision avoidance)_ and then you can access the API directly.

On the board side of the interface, it’s likely just TTL/GPIO/UART, I2C, or SPI serial data, right?

I’ve long ago misplaced the relevant patent number(s), but with a bit of searching you might also find illuminating details in the patent description and drawings.

- - - -
**Update:** I found what appears to be the relevant patents, thanks to [this page](http://www.bloominglabs.org/index.php/Water_Heater_Monitoring_and_Setback_Control).

The following patent description both contains a schematic _(including a parts list!)_ for an EnviraCOM transceiver and also the voltage levels\* _(scroll down to “Table 1”)_:

- https://patents.google.com/patent/US7769932

\* _(which for the higher of the two signals appears to be about 12-18 VDC … basically whatever the circuit gets by using a half- or full-wave rectifier on 24VAC)_

This may be for the EnviroCOM 1.1 or 2.0 protocol versions, as it says it uses the AC signal crossings as a synchronization source to get the data rate up to about 1 kb/s from 120 bps.

- - - -
_One final note, as I haven’t seen this mentioned previously:_

- While _EnviraCOM_ is similar to the CAN _(Controller Area Network)_ bus protocol, it uses a single data conductor instead of the differential pair commonly used in other CAN-type busses; it also appears to be an evolution _(or at least a re-branding)_ of an older Honeywell multi-device 3-wire communication bus which they referred to as **”low end subsystem topology (LESST)“**.

<details><summary>More on ‘LESST’:</summary>

_Found on a [LinkedIn profile](https://www.linkedin.com/in/ravikurupati):_

> Software Engineer: Honeywell Technology Solutions Pvt. Ltd. (Jan 1998 – Sep 1999)
> - Developed firmware for Honeywell’s Home Controller’s I/O board which is the main interface to all the different home networks (LESST, CE, Serial). Involves writing C and Assembly code for PIC17C756 16 bit high-end controller.
> - Designed and developed firmware for a smart temperature sensor that’s part of the Honeywell’s LESST home network. Involves writing assembly code for 68705KJ1 8 bit low-end micro controllers. 
> - Developed a PC based telephone interface module to simulator a voice-activated system. This gives remote access to the user to control and monitor his house. It talks to the telephone and the devices sitting on the LESST bus
> - Designed and developed network monitoring tools like RF analyzer and LESST monitor to test the firmware on the embedded devices.

**Comment by jayepolo** [comment]
I would love to put a scope on the pins.  I have a boiler that talks
EnviraCOM.  It is a serial out.  I presume it is not easy to directly
connect an ESP8266, but that is the end state I would like to establish.
Ideally to create a board that interfaces to EnviroCOM and bridges to
Wifi.  I need a cheap scope, or better info on the physical layer of the
serial connection...

**Comment by jgrisham** [comment]
Connecting to the external pins of the boiler interface would indeed be more than trivial work, due to the voltage levels and the need to monitor the AC lines for timing synchronization - and that is just for passive monitoring. That could _probably_ be accomplished with an Arduino or Raspberry Pi and only a few external components _(mostly for isolation)_, but may be quite beyond the capabilities of a lone ESP8266.

Transmitting would be even more difficult.

Perhaps someone would be willing to put together _(e.g. on a solderless breadboard)_ the circuit described in the patent I linked to above to see if a simple interface would be possible _(that would then go **between** the boiler interface and the ESP8266)_. 
- - - -
**If you can find a photo or schematic of the _inside_ of a boiler interface of the same model as yours**, then it would be easier to determine if _bypassing_ the external interface is feasible. If there is already circuitry there doing all the hard work, it might not be necessary to reverse-engineer and then rebuild that driver circuitry if it is possible to tap in upstream where there is a plain TTL or CMOS signal. There may even be debug pins / pads present… _if_ that were the case then using an ESP8266 _would_ be trivial.

There were _(5-10 years ago?)_ reports of the _official Honeywell serial adapter_ overheating… which makes me even more wary of trying to interface _directly_ with the 3-wire EnviraCOM bus… but then again, I’m old and tired. 😉

**Comment by jgrisham** [comment]
@jaypolo FYI, I edited [my original post](https://github.com/roy-bentley/enviracom/issues/1#issuecomment-992163503) after you likely received your e-mail notification…

… that post now contains a link to a patent, [US7769932](https://patents.google.com/patent/US7769932), that describes the physical layer _(among others)_ of the Honeywell EnviraCOM protocol\*, including the voltages _(e.g. data 1 is 12-18 VDC, and normally should be limited to <= 17 VDC)_, bus arbitration methods and timing, and even a reference interface circuit schematic including a detailed parts list.

\* _(I did not determine if that specific patent was describing version 1.1 or version 2.0, though. If I had to guess, v1.0 is a re-branding of ‘Honeywell LESST’ and 2.0 might be the wireless version used by the 2-wire thermostats… if both of those assumptions were true, then that patent likely describes EnviraCOM v1.1.)_

**Comment by jayepolo** [comment]
Oh very  interesting to explore the patent - thank you for sharing that!
This is super helpful and send me down the rabbit hole for a while!!

My MVP use case is a single connection directly from an ESP to my boiler
controller.  No thermostats, just an ESP connected to a single EnviraCOM
device.  Hopefully that will simplify both hardware and software.  At
first, simply by only receiving anything automatically initiated by the
boiler.   Eventually, I would like to request data and if I do end up
initiating transmissions, I might actually be able to tolerate and recover
from any collisions that could potentially happen.

Interesting.  Should be fun to dig into this...  If you have any other
suggestions on getting started, I am all ears, thanks.

**Comment by DiskDoctor** [comment]
I found EnviraCOM document that has more details on the serial adapter and the various message types.

I would like to add an outdoor thermostat to my L7224 triple aquastat that supports E-COM. If someone can think of a way to make an interface that connects to the 3-wire bus for a I2C, Dallas one-wire, or even analog sensor, that would be great. 

Arduino/ESP based software could use such an interface board and allow all sorts of functions.

A reminder that 24v is AC voltage on HVAC systems. 

[EnviroCom Bus 69-2644.pdf](https://github.com/roy-bentley/enviracom/files/7819928/EnviroCom.Bus.69-2644.pdf)
.

**Comment by jayepolo** [comment]
Sounds interesting.  But I cannot seem to navigate to that document on
Github.  Could you attach it here?

**Comment by jayepolo** [comment]
Ignore that, I do have it.  Thx.

**Comment by sn3kyJ3di** [comment]
@jayepolo hello, I was just reading through the project as I have been considering something similar, did you ever make any headway?

**Comment by jayepolo** [comment]
@sn3kyJ3di  no, unfortunately I have not.  Too many irons in the fire.  What I did do is find an inexpensive EnviraCom alarm module on eBay.  I've got it mounted to the boiler but have not yet wired in.  It will monitor the bus and close a relay in the event of an error message being transmitted over the bus.

Instead, I have prioritized capturing various inputs that can otherwise give me some insights into how my system is working.  That data all gets transmitted with an ESP32 and recorded in the cloud.

I do still intend to connect the EnviraCom alarm next, which can proactively alert me to issues.  And ultimately, I would still like to be able to interact over the EnviraCom bus.    

What are you hoping to build?

**Comment by sn3kyJ3di** [comment]
I would love to be able to connect an ESP32 module directly to the com terminals on my hot water heater gas valve controller so I can track gas usage and "setback" at times of no demand.  My gas bill has doubled in 12 months.  I'm also a IOT/automation enthusiast and have my water and electric metered to my Home Assistant instance, gas is just one of the missing pieces.  It seems like you can spend 150-200$ to get it working but where is the fun in that?

**Comment by jayepolo** [comment]
I suspect you could monitor operation pretty easily.  Maybe not via the EnviraCom, but simply by detecting the flame with a light sensor.  But I am not sure how to control setback.  Is the hot water temperature setpoint a digital setting or is it an analog dial?  (Mine is the latter).  I am happy to point you to my ESP code - it is very very simple.  The real task is in interfacing with your system...

**Comment by 5310H** [comment]
Any update on this?  Also want to interface my gas water heater into HA.

**Comment by jayepolo** [comment]
No activity on my part.  Too many irons in the fire...
